File: etc/java/BaseXClient.java
import java.io.BufferedInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

/**
 * Language Binding for BaseX.
 * Works with BaseX 6.1.9 and later
 * Documentation: http://basex.org/api
 *
 * (C) Workgroup DBIS, University of Konstanz 2005-10, ISC License
 */
public class BaseXClient {
  /** Output stream. */
  final OutputStream out;
  /** Socket. */
  private final Socket socket;
  /** Cache. */
  private final BufferedInputStream in;
  /** Command info. */
  private String info;

  /**
   * Constructor.
   * @param host server name
   * @param port server port
   * @param usern user name
   * @param pw password
   * @throws IOException Exception
   */
  public BaseXClient(final String host, final int port, final String usern,
      final String pw) throws IOException {

    socket = new Socket();
    socket.connect(new InetSocketAddress(host, port), 5000);
    in = new BufferedInputStream(socket.getInputStream());
    out = socket.getOutputStream();

    // receive timestamp
    final String ts = receive();
    // send {Username}0 and hashed {Password/Timestamp}0
    send(usern);
    send(md5(md5(pw) + ts));

    // receive success flag
    if(!ok()) throw new IOException("Access denied.");
  }

  /**
   * Executes a command and serializes the result to an output stream.
   * @param cmd command
   * @param o output stream
   * @throws IOException Exception
   */
  public void execute(final String cmd, final OutputStream o)
      throws IOException {
    // send {Command}0
    send(cmd);
    receive(o);
    info = receive();
    if(!ok()) throw new IOException(info);
  }

  /**
   * Executes a command and returns the result.
   * @param cmd command
   * @return result
   * @throws IOException Exception
   */
  public String execute(final String cmd) throws IOException {
    final OutputStream os = new ByteArrayOutputStream();
    execute(cmd, os);
    return os.toString();
  }

  /**
   * Creates a query object.
   * @param query query string
   * @return query
   * @throws IOException Exception
   */
  public Query query(final String query) throws IOException {
    return new Query(query);
  }

  /**
   * Creates a database.
   * @param name name of database
   * @param input xml input
   * @throws IOException I/O exception
   */
  public void create(final String name, final InputStream input)
      throws IOException {
    // send 3 to mark start of query execution, and {Query}0 as query string
    out.write(3);
    send(name);
    int l;
    while((l = input.read()) != -1) out.write(l);
    out.write(0);
    info = receive();
    if(!ok()) throw new IOException(info);
  }

  /**
   * Returns command information.
   * @return string info
   */
  public String info() {
    return info;
  }

  /**
   * Closes the session.
   * @throws IOException Exception
   */
  public void close() throws IOException {
    send("exit");
    out.flush();
    socket.close();
  }

  /**
   * Checks the next success flag.
   * @return value of check
   * @throws IOException Exception
   */
  boolean ok() throws IOException {
    out.flush();
    return in.read() == 0;
  }

  /**
   * Returns the next received string.
   * @return String result or info
   * @throws IOException I/O exception
   */
  String receive() throws IOException {
    final OutputStream os = new ByteArrayOutputStream();
    receive(os);
    return os.toString();
  }

  /**
   * Sends a string to the server.
   * @param s string to be sent
   * @throws IOException I/O exception
   */
  void send(final String s) throws IOException {
    out.write((s + '\0').getBytes("UTF8"));
  }

  /**
   * Receives a string and writes it to the specified output stream.
   * @param o output stream
   * @throws IOException I/O exception
   */
  private void receive(final OutputStream o) throws IOException {
    out.flush();
    while(true) {
      final int b = in.read();
      if(b == 0 || b == -1) break;
      o.write(b);
    }
  }

  /**
   * Returns an MD5 hash.
   * @param pw String
   * @return String
   */
  private static String md5(final String pw) {
    final StringBuilder sb = new StringBuilder();
    try {
      final MessageDigest md = MessageDigest.getInstance("MD5");
      md.update(pw.getBytes());
      for(final byte b : md.digest()) {
        final String s = Integer.toHexString(b & 0xFF);
        if(s.length() == 1) sb.append('0');
        sb.append(s);
      }
    } catch(final NoSuchAlgorithmException ex) {
      // should not occur
      ex.printStackTrace();
    }
    return sb.toString();
  }

  /**
   * Inner class for iterative query execution.
   */
  class Query {
    /** Query id. */
    private final String id;
    /** Next result item. */
    private String next;

    /**
     * Standard constructor.
     * @param query query string
     * @throws IOException I/O exception
     */
    public Query(final String query) throws IOException {
      // send 0 to mark start of query execution, and {Query}0 as query string
      out.write(0);
      send(query);
      id = receive();
      if(!ok()) throw new IOException(receive());
    }

    /**
     * Checks for the next item.
     * @return value of check
     * @throws IOException I/O exception
     */
    public boolean more() throws IOException {
      // send 1 to get next result item, and {ID}0 for identification
      out.write(1);
      send(id);
      next = receive();
      if(!ok()) throw new IOException(receive());
      return next.length() != 0;
    }

    /**
     * Returns the next item.
     * @return item string
     */
    public String next() {
      return next;
    }

    /**
     * Closes the query.
     * @throws IOException I/O exception
     */
    public void close() throws IOException {
      // send 2 to mark end of query execution, and {ID}0 for identification
      out.write(2);
      send(id);
      out.flush();
    }
  }
}

File: etc/php/BaseXClient.php
<?php
/*
 * Language Binding for BaseX.
 * Works with BaseX 6.3.1 and later
 * Documentation: http://basex.org/api
 * 
 * (C) BaseX Team 2005-11, ISC License
 */
class Session {
  /* Class variables.*/
  var $socket, $info, $buffer, $bpos, $bsize;

  /* see readme.txt */
  function __construct($h, $p, $user, $pw) {
    // create server connection
    $this->socket = socket_create(AF_INET, SOCK_STREAM, SOL_TCP);
    if(!socket_connect($this->socket, $h, $p)) {
      throw new Exception("Can't communicate with server.");
    }

    // receive timestamp
    $ts = $this->receive();

    // send username and hashed password/timestamp
    $md5 = hash("md5", hash("md5", $pw).$ts);
    socket_write($this->socket, "$user\0$md5\0");

    // receives success flag
    if(socket_read($this->socket, 1) != "\0") {
      throw new Exception("Access denied.");
    }
  }

  /* see readme.txt */
  public function execute($com) {
    // send command to server
    socket_write($this->socket, "$com\0");

    // receive result
    $result = $this->receive();
    $this->info = $this->receive();
    if($this->ok() != True) {
      throw new Exception($this->info);
    }
    return $result;
  }

  /* see readme.txt */
  public function info() {
    return $this->info;
  }

  /* see readme.txt */
  public function close() {
    socket_write($this->socket, "exit\0");
    socket_close($this->socket);
  }

  /* Initializes the byte transfer */
  private function init() {
    $this->bpos = 0;
    $this->bsize = 0;
  }

  /* Receives a string from the socket. */
  public function receive() {
    $com = "";
    while(($d = $this->read()) != "\0") {
      $com .= $d;
    }
    return $com;
  }

  /* Returns a single byte from the socket. */
  private function read() {
    if($this->bpos == $this->bsize) {
      $this->bsize = socket_recv($this->socket, $this->buffer, 4096, 0);
      $this->bpos = 0;
    }
    return $this->buffer[$this->bpos++];
  }
  
  /* Returns the query object.*/
  public function query($q) {
    return new Query($this, $q);
  }
  
  /* Sends the str. */
  public function send($str) {
    socket_write($this->socket, "$str\0");
  }
  
  /* Returns success check. */
  public function ok() {
    return $this->read() == "\0";
  }
  
  /* Returns the result. */
  public function receive() {
    $this->init();
    return $this->receive();
  }
}

class Query {
  /* Class variables.*/
  var $session, $id, $open, $next;
 
  /* see readme.txt */
  function __construct($s, $q) {
    $this->session = $s;
    $this->id = $this->exec("\0", $q);
  }
  
  /* see readme.txt */
  function init() {
    return $this->exec("\4", $this->id);
  }
  
  /* see readme.txt */
  function bind($name, $value) {
    $this->exec("\3", "$this->id\0$name\0$value\0");
  }

  /* see readme.txt */
  public function more() {
    $this->next = $this->exec("\1", $this->id);
    return strlen($this->next) > 0; 
  }

  /* see readme.txt */
  public function next() {
    return $this->next;
  }
  
  /* see readme.txt */
  public function execute() {
    return $this->exec("\5", $this->id);
  }
  
  /* see readme.txt */
  public function info() {
    return $this->exec("\6", $this->id);
  }
  
  /* see readme.txt */
  public function close() {
  	return $this->exec("\2", $this->id);   
  }
  
  /* see readme.txt */
  public function exec($cmd, $arg) {
  	$this->session->send("$cmd$arg");
  	$s = $this->session->receive();
  	if($this->session->ok() != True) {
  	  throw new Exception($this->session->receive());
  	}
  	return $s;
  }
}
?>
